### ISSUE TITLE
Return after halt() being ignored can be confusing

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#24335]
An impact of https://github.com/chapel-lang/chapel/pull/23309 that I'm only now understanding is that we can no longer use a `return` in an abstract base class method to define the return type of that method.  While this is probably acceptable, the current error doesn't particularly help a user determine what's wrong.

As an example, consider the following code:

```chapel
class Collection {
  proc size() {
    halt("This is an abstract method and shouldn't be called");
    return -1;
  }
}

class List: Collection {
  override proc size() {
    return 10;
  }
}

class Set: Collection {
  override proc size() {
    return 12;
  }
}

var c1, c2: Collection;
c1 = new List();
c2 = new Set();

writeln(c1.size());
writeln(c2.size());
```

Prior to this change, the return-after-halt pattern was sufficient to define that `size()` should return an `int`.  With the change, we now get the error:

```
1: In module 'classHier':
24: error: illegal use of function that does not return a value: 'size'
```

Yet, inspecting the function in question quickly and seeing the `return`, it looks as though it returns a value, which can cause confusion (at least, it did for me).  If the error were able to say that the routine didn't return anything due to the halt, or simply complained about the return following the halt() being pointless / a no-op, that seems like it would clarify the situation.

**Comment by vasslitvinov** [comment]
#20673 raises other related questions.

We should at the very least add a clarification to the spec [here](https://chapel-lang.org/docs/language/spec/procedures.html#implicit-return-types).

**Comment by mppf** [comment]
IMO for the specific pattern here, we could introduce a library routine to call to indicate a method is "pure virtual" / "abstract" and call that instead of `halt`. Then, the compiler can be adjusted to not consider this to make the `return` dead code. Of course we can consider other ways to indicate something is pure virtual, such as an attribute or keyword. See also #8566 and #15271.

I would also be open to an approach that uses such top-level `return`s for type inference of routine return types even if there is `halt` before it, possible as an exception to the general dead code / param folding rules.

**Comment by vasslitvinov** [comment]
In my view, the expectation that code after a `halt()` will do anything is legacy. For today's Chapel, I see it "normal" to declare the return type explicitly rather than relying on return-after-halt.

Given the importance of abstract classes and pure virtual methods it makes sense to introduce language support for them. A library routine, an attribute, or some syntax (cf. `init tnis`) are possibilities.

That said, if others find return-after-halt valuable, I can live with it.

**Comment by bradcray** [comment]
> Of course we can consider other ways to indicate something is pure virtual, such as an attribute or keyword. See also https://github.com/chapel-lang/chapel/issues/8566 and https://github.com/chapel-lang/chapel/issues/15271.

> Given the importance of abstract classes and pure virtual methods it makes sense to introduce language support for them.

Good points and I would also prefer language support for pure virtual methods than continuing to use this pattern to fake it.

**Comment by dlongnecke-cray** [comment]
I agree with Vass and think it would be reasonable for the language to detect and warn against this pattern, perhaps suggesting you annotate a return type instead. I also think it would be reasonable to try and warn for all unreachable code that is not being deliberately param folded away.
